bracketcolors/src/bracketcolors.cc

```diff
@@ -0,0 +1,1303 @@
+        g_source_remove(computeTimeoutID);
+        computeTimeoutID = 0;
+    }
+
+    if (drawTimeoutID > 0) {
+        g_source_remove(drawTimeoutID);
+        drawTimeoutID = 0;
+    }
+}
+
+
+
+// -----------------------------------------------------------------------------
+    void BracketColorsData::RemoveFromQueues(BracketMap::Index index)
+
+/*
+
+----------------------------------------------------------------------------- */
+{
+    {
+        auto it = recomputeIndicies.find(index);
+        if (it != recomputeIndicies.end()) {
+            recomputeIndicies.erase(it);
+        }
+    }
+    {
+        auto it = redrawIndicies.find(index);
+        if (it != redrawIndicies.end()) {
+            redrawIndicies.erase(it);
+        }
+    }
+}
+
+
+// -----------------------------------------------------------------------------
+    static gboolean utils_is_dark(guint32 color)
+
+/*
+
+----------------------------------------------------------------------------- */
+{
+    guint8 b = color >> 16;
+    guint8 g = color >> 8;
+    guint8 r = color;
+
+    // https://stackoverflow.com/questions/596216/formula-to-determine-perceived-brightness-of-rgb-color
+    guint8 y  = ((r << 1) + r + (g << 2) + b) >> 3;
+
+    if (y < 125) {
+        return TRUE;
+    }
+
+    return FALSE;
+}
+
+
+
+// -----------------------------------------------------------------------------
+    static gboolean utils_parse_color(
+        const gchar *spec,
+        GdkColor *color
+    )
+/*
+
+----------------------------------------------------------------------------- */
+{
+    gchar buf[64] = {0};
+
+    g_return_val_if_fail(spec != NULL, -1);
+
+    if (spec[0] == '0' && (spec[1] == 'x' || spec[1] == 'X'))
+    {
+        /* convert to # format for GDK to understand it */
+        buf[0] = '#';
+        strncpy(buf + 1, spec + 2, sizeof(buf) - 2);
+        spec = buf;
+    }
+
+    return gdk_color_parse(spec, color);
+}
+
+
+
+// -----------------------------------------------------------------------------
+    static gint utils_color_to_bgr(const GdkColor *c)
+/*
+
+----------------------------------------------------------------------------- */
+{
+    g_return_val_if_fail(c != NULL, -1);
+    return (c->red / 256) | ((c->green / 256) << 8) | ((c->blue / 256) << 16);
+}
+
+
+
+// -----------------------------------------------------------------------------
+    static gint utils_parse_color_to_bgr(const gchar *spec)
+/*
+
+----------------------------------------------------------------------------- */
+{
+    GdkColor color;
+    if (utils_parse_color(spec, &color)) {
+        return utils_color_to_bgr(&color);
+    }
+    else {
+        return -1;
+    }
+}
+
+
+
+// -----------------------------------------------------------------------------
+    static gchar char_at(ScintillaObject *sci, gint pos)
+/*
+
+----------------------------------------------------------------------------- */
+{
+    return sci_get_char_at(sci, pos);
+}
```

Comment:
Just curious: why the wrapper?

Reply:
> Just curious: why the wrapper?

The [documented return type of `SCI_GETCHARAT`] is `int`, presumably for Unicode characters up to 32 bits in length. Squeezing that into an 8-bit(?) `(g)char` needs an explicit cast.

The (implied) point that just using the returned `(g)int` without a cast is worth considering. At the same time, the `pos` argument might be safer as the (platform independent) [`Sci_Position` type]. The API structures that can *only* index 32-bit positions are [slated for eventual deprecation].

[documented return type of `SCI_GETCHARAT`]: https://www.scintilla.org/ScintillaDoc.html#SCI_GETCHARAT
[`Sci_Position` type]: https://github.com/geany/geany/blob/8a9e9be88545f83c91d378d6662790efa38959d4/scintilla/include/Sci_Position.h#L14-L15
[slated for eventual deprecation]: https://groups.google.com/g/scintilla-interest/c/mPLwYdC0-FE#:~:text=SCI_GETTEXTRANGE%20will%20be%20deprecated%20at%20some%20point%20(likely%206.0)

Reply:
> The [documented return type of `SCI_GETCHARAT`](https://www.scintilla.org/ScintillaDoc.html#SCI_GETCHARAT) is `int`, presumably for Unicode characters up to 32 bits in length.

No, that's probably either because `int` is "the" basic numerical type in C, or because of other historical reasons (and technically it is returned as a `sptr_t`). In practice, the returned value comes from `Document::ChatAt()` which returns a `char`. So if if want to be picky, it's actually a `char` stored into a `sptr_t`.

> Squeezing that into an 8-bit(?) `(g)char` needs an explicit cast.

Well, yes, but… actually no: this does not actually wrap `SCI_GETCHARAT`, but Geany's own wrapper, `sci_get_char_at()` *which has the exact same signature*.

Reply:
When I was writing this, I was using the `autoclose` plugin as a template [which has this wrapper as well](https://github.com/geany/geany-plugins/blob/dc6980e5df24434597b37a5a7064837b3f5158e8/autoclose/src/autoclose.c#L93). I suspect the authors of `autoclose` used the wrapper for convenience since it's used a lot, I don't really need it as bad to be honest.